cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/shortestpath/SingleSourceShortestPath.scala

```diff
@@ -0,0 +1,44 @@
+package com.twitter.cassovary.algorithms.shortestpath
+
+import com.twitter.cassovary.graph.{Node, GraphDir, BreadthFirstTraverser, DirectedGraph}
+
+case class SingleSourceShortestPath(graph: DirectedGraph[Node], source: Int) extends ShortestPath {
+
+  val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(source))
+  private val (nodes, depths)  = (bfs.toList, bfs.allDepths)
+  private val previous = nodes.map { n =>
+    n.id -> n.inboundNodes().filter { in => depths(in) == depths(n.id) - 1}
+  }.toMap
+
+  private def walkOnGraph(currStack: Stack, currTop: Int): (Stack, Int, Path) = {
+    val node  = currStack(currTop)(0)
+    val index = currStack(currTop)(1)
+
+    val path = if (node == source)
+      currStack.take(currTop + 1).reverse.map { k => k(0) }.toSeq
+    else
+      Seq.empty[Int]
+
+    val (stack, top) = if (previous(node).length > index) {
+      val newStack = if (currTop + 1 == currStack.length)
+        currStack :+ Seq(previous(node)(index), 0)
+      else
+        currStack.zipWithIndex.map { case (seq, in) => if (in == currTop + 1) Seq(previous(node)(index), 0) else seq}
+      (newStack, currTop + 1)
+    }
+    else {
+      val newStack = currStack.zipWithIndex.map { case (seq, in) => if (in == currTop - 1) Seq(seq(0), seq(1) + 1) else seq}
+      (newStack, currTop - 1)
+    }
+    (stack, top, path)
+  }
+
+  def shortestPaths(target: Int): Paths = {
+    lazy val stream: Stream[(Stack, Int, Path)] =
+      (List(Seq(target, 0)), 0, Seq.empty[Int]) #:: stream.map { case (s, t, p) => walkOnGraph(s,t) }
+
+    stream.takeWhile{ case (_, t, _) => t >= 0}.toList
+      .filter{ case (_, _, p) => p.nonEmpty }
+      .map{ _._3 }.toSeq
```

Comment:
Talking just about walkOnGraph(), I don’t understand it fully but I think what you are doing is retracing back from target node and keeping track of ancestor in ‘index’. My first question is why you are tracing backwards (which relies on ‘previous’) and not moving forwards out of source node along its outneighbors, in which case this can be much simpler as a variation of BFS where you keep track of parent of node as well as the shortest path length. You enqueue the node only if the current depth is equal or less than the shortest path length value currently stored in the node.
Taking g6 as an example, you would start from source 10 say, enqueue 11,12,13 and keep 10 as ancestor for each of them as well as record a value of ‘1’ for the shortest path length to them. Then consider node 11 which has two children 12 and 14:

You will enqueue 14 (together with ancestor=11, path length=2) because no shortest path length has been stored with 14 so far.

You will not enqueue 12 but already a shorter path length of 1 has been stored.

This can then be coded just like BFS as an extension of queuebased traverser in Traverser.scala which would be the preferred way of doing the all shortest paths calculation.


> On Mar 3, 2015, at 7:57 AM, Ben McKown <notifications@github.com> wrote:
> 
> In cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/shortestpath/SingleSourceShortestPath.scala <https://github.com/twitter/cassovary/pull/166#discussion_r25698526>:
> 
> > +      (newStack, currTop + 1)
> > +    }
> > +    else {
> > +      val newStack = currStack.zipWithIndex.map { case (seq, in) => if (in == currTop - 1) Seq(seq(0), seq(1) + 1) else seq}
> > +      (newStack, currTop - 1)
> > +    }
> > +    (stack, top, path)
> > +  }
> > +
> > +  def shortestPaths(target: Int): Paths = {
> > +    lazy val stream: Stream[(Stack, Int, Path)] =
> > +      (List(Seq(target, 0)), 0, Seq.empty[Int]) #:: stream.map { case (s, t, p) => walkOnGraph(s,t) }
> > +
> > +    stream.takeWhile{ case (_, t, _) => t >= 0}.toList
> > +      .filter{ case (_, _, p) => p.nonEmpty }
> > +      .map{ _._3 }.toSeq
> walkonGraph() is probably the most opaque piece of code here. Essentially, I am trying to efficiently determine all shortest paths from node i to node j--not just a shortest path from node i to node j.
> 
> Here is the output from the stack manipulation when determining the paths from node 10 to node 14 in TestGraphs.g6. The result we are after is a set of paths [(10,11,14), (10,12,14), (10,13,14)]. The first line is the state of the current stack, the second line is the value of current "top"--which is essentially how far we have travelled along a given path (which could include traveling backward).
> 
> List(List(14, 0))
> 0
> List(List(14, 0), List(11, 0))
> 1
> List(List(14, 0), List(11, 0), List(10, 0))
> 2
> List(List(14, 0), List(11, 1), List(10, 0))
> 1
> List(List(14, 1), List(11, 1), List(10, 0))
> 0
> List(List(14, 1), List(12, 0), List(10, 0))
> 1
> List(List(14, 1), List(12, 0), List(10, 0))
> 2
> List(List(14, 1), List(12, 1), List(10, 0))
> 1
> List(List(14, 2), List(12, 1), List(10, 0))
> 0
> List(List(14, 2), List(13, 0), List(10, 0))
> 1
> List(List(14, 2), List(13, 0), List(10, 0))
> 2
> List(List(14, 2), List(13, 1), List(10, 0))
> 1
> List(List(14, 3), List(13, 1), List(10, 0))
> 0
> So, we keep track of a list of tuples(node, index). Index is probably a bad name for the second part as it really keeps track of the number of shortest paths that pass through that given node. So, we keep walking starting at target and ending at source and keep visiting every node. And at each node we visit, we ask "Is there another node that will lead me to source?" If so, we follow that path, if not, we walk backward further until we do reach a node that could lead to another shortest path.
> 
> Just to be illustrative, here is the output for all shortest paths from node 10 to node 15.
> 
> List(List(15, 0))
> 0
> List(List(15, 0), List(14, 0))
> 1
> List(List(15, 0), List(14, 0), List(11, 0))
> 2
> List(List(15, 0), List(14, 0), List(11, 0), List(10, 0))
> 3
> List(List(15, 0), List(14, 0), List(11, 1), List(10, 0))
> 2
> List(List(15, 0), List(14, 1), List(11, 1), List(10, 0))
> 1
> List(List(15, 0), List(14, 1), List(12, 0), List(10, 0))
> 2
> List(List(15, 0), List(14, 1), List(12, 0), List(10, 0))
> 3
> List(List(15, 0), List(14, 1), List(12, 1), List(10, 0))
> 2
> List(List(15, 0), List(14, 2), List(12, 1), List(10, 0))
> 1
> List(List(15, 0), List(14, 2), List(13, 0), List(10, 0))
> 2
> List(List(15, 0), List(14, 2), List(13, 0), List(10, 0))
> 3
> List(List(15, 0), List(14, 2), List(13, 1), List(10, 0))
> 2
> List(List(15, 0), List(14, 3), List(13, 1), List(10, 0))
> 1
> List(List(15, 1), List(14, 3), List(13, 1), List(10, 0))
> 0
> As you can see here, we only "walk" back to 15 once since we only need to walk back to 14 before we find another node that will lead to a shortest path. I often find this output to be the most helpful in understanding why and how this algorithm works. As per usual, I followed the example from networkx for shortest paths. I will try to find the source.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/twitter/cassovary/pull/166/files#r25698526>.
>